library/std/src/sys_common/once/generic.rs

```diff
@@ -0,0 +1,280 @@
+//   it needs to see the changes to drop the `Waiter` struct correctly.
+// * There is one place where the two atomics `Once.state_and_queue` and
+//   `Waiter.signaled` come together, and might be reordered by the compiler or
+//   processor. Because both use acquire ordering such a reordering is not
+//   allowed, so no need for `SeqCst`.
+
+use crate::cell::{Cell, UnsafeCell};
+use crate::mem::ManuallyDrop;
+use crate::ptr;
+use crate::sync as public;
+use crate::sync::atomic::{
+    AtomicBool, AtomicPtr,
+    Ordering::{AcqRel, Acquire, Release},
+};
+use crate::thread::{self, Thread};
+
+struct Masked;
+
+pub struct Once {
+    state_and_queue: AtomicPtr<Masked>,
+}
+
+pub struct OnceState {
+    poisoned: bool,
+    set_state_on_drop_to: Cell<*mut Masked>,
+}
+
+// Four states that a Once can be in, encoded into the lower bits of
+// `state_and_queue` in the Once structure.
+const INCOMPLETE: usize = 0x0;
+const POISONED: usize = 0x1;
+const RUNNING: usize = 0x2;
+const COMPLETE: usize = 0x3;
+
+// Mask to learn about the state. All other bits are the queue of waiters if
+// this is in the RUNNING state.
+const STATE_MASK: usize = 0x3;
+
+#[inline]
+fn to_state(waiter: &Waiter) -> *mut Masked {
+    (waiter as *const Waiter as *const Masked as *mut Masked).map_addr(|addr| addr | RUNNING)
+}
+
+#[inline]
+fn to_waiter(state: *mut Masked) -> *const Waiter {
+    state.map_addr(|addr| addr & !STATE_MASK) as *mut Waiter
+}
+
+// Representation of a node in the linked list of waiters, used while in the
+// RUNNING state.
+#[repr(align(4))] // Ensure the two lower bits are free to use as state bits.
+struct Waiter {
+    // Use `ManuallyDrop` instead of `Option`, as trying to wake the waiter from
+    // multiple threads is UB anyways.
+    thread: UnsafeCell<ManuallyDrop<Thread>>,
+    signaled: AtomicBool,
+    next: *const Waiter,
+}
+
+unsafe impl Send for Waiter {}
+unsafe impl Sync for Waiter {}
+
+// Head of a linked list of waiters.
+// Every node is a struct on the stack of a waiting thread.
+// Will wake up the waiters when it gets dropped, i.e. also on panic.
+struct WaiterQueue<'a> {
+    state_and_queue: &'a AtomicPtr<Masked>,
+    set_state_on_drop_to: *mut Masked,
+}
+
+impl Once {
+    #[inline]
+    pub const fn new() -> Once {
+        Once { state_and_queue: AtomicPtr::new(ptr::invalid_mut(INCOMPLETE)) }
+    }
+
+    #[inline]
+    pub fn is_completed(&self) -> bool {
+        // An `Acquire` load is enough because that makes all the initialization
+        // operations visible to us, and, this being a fast path, weaker
+        // ordering helps with performance. This `Acquire` synchronizes with
+        // `Release` operations on the slow path.
+        self.state_and_queue.load(Acquire).addr() == COMPLETE
+    }
+
+    // This is a non-generic function to reduce the monomorphization cost of
+    // using `call_once` (this isn't exactly a trivial or small implementation).
```

Comment:
I've copied this from the old implementation, but I'm not sure whether that's really necessary?